### DISCUSSION TITLE
Why Differencer always returned fitted data when indices are the same?

### DISCUSSION THREAD

**Discussion by ilkersigirci** [alan-turing-institute/sktime#3277] [Q&A] [answered]
Differencer transformer returns fitted data as the [source code](https://github.com/alan-turing-institute/sktime/blob/main/sktime/transformations/series/difference.py#L268) says. I would expect it to be dynamically calculated with the given data in the `inverse_transform` case. I don't if this is a bug or intended behavior, hence I opened a discussion first. If requested, I can also open bug report. Below codes show the difference between default sktime approach and my use-case.

```python
# Sktime approach
from sktime.datasets import load_airline
from sktime.transformations.series.difference import Differencer

data = load_airline()

# y and y_reconstructed completely the same
transformer = Differencer(lags=1, na_handling="fill_zero")
y_transform = transformer.fit_transform(data)
y_transform += 1000
y_reconstructed = transformer.inverse_transform(y_transform)
```

```python
# My use-case
# Code is adapted from https://machinelearningmastery.com/remove-trends-seasonality-difference-transform-python/
from typing import Union
from sktime.datasets import load_airline

data = load_airline()

def difference(dataset, lags=1):
	diff = list()
	for i in range(lags, len(dataset)):
		value = dataset[i] - dataset[i - lags]
		diff.append(value)
	return pd.Series(diff)
 

def inverse_difference(last_ob, value):
	return value + last_ob

y_transform = difference(data)
y_transform += 1000

# y and y_reconstructed aren't the same
y_reconstructed = [inverse_difference(data[i], y_transform[i]) for i in range(len(y_transform))]
```

**Versions**
<details>
System:
    python: 3.7.13 (default, Apr 24 2022, 01:04:09)  [GCC 7.5.0]
executable: /usr/bin/python3
   machine: Linux-5.4.188+-x86_64-with-Ubuntu-18.04-bionic

Python dependencies:
          pip: 21.1.3
   setuptools: 57.4.0
      sklearn: 1.0.2
       sktime: 0.13.1
  statsmodels: 0.13.2
        numpy: 1.21.6
        scipy: 1.7.3
       pandas: 1.3.5
   matplotlib: 3.2.2
       joblib: 1.1.0
        numba: 0.56.0
     pmdarima: 1.8.5
      tsfresh: None

</details>

**Comment by fkiraly** [comment]
I think this is a very interesting and very subtle question, much appreciated!

I think there are two user expectations that are clashing here:
* neither `transform` nor `inverse_transform` rely on data seen in `fit`, becaus we can just compute differences
* `inverse_transform` and `transform` are inverse to each other

and the design principle that all tansformers have the same interface (i.e., no extra `value` in the `inverse_transform`)

The key problem is that taking n-th differences is *not an injective operation*.

Example:

`y_orig = [1, 2, 5]`

`difference(y_orig) = [1, 3]`

`naive_inverse_difference(difference(y_orig)) = [0, 1, 4]` (if we choose `value` to be a default 0)

A common use case is using `Differencer` as part of the `TransformedTargetForecaster`, so combined with the above, we need to memorize some data (the last indices of the training data) to ensure the differencing is inverted correctly.

Re your example: I do not fully understand what you are trying to say, could you show an example with the `sktime` differencer, with actual and expected, plus a rationale why/what you expect?

**Comment by ilkersigirci** [comment]
@fkiraly Sorry for the late answer. Assume that I want to log&diff y and X and use ARIMA model as forecaster. When I tried to get in_sample prediction of the pipeline, I got exactly original y data instead of arima's in_sample prediction. Note that this is not the case when using just the model not the pipeline with the transformer. I think this happens because Differencer returns its stored data for the same indices. 

I believe this is a general bug for any forecasting pipeline that uses Differencer and forecaster whose able to do in_sample prediction together.

```python
import numpy as np
from sktime.forecasting.compose import (
    ForecastingPipeline,
    TransformedTargetForecaster,
)
from sktime.transformations.compose import TransformerPipeline
from sktime.transformations.series.compose import OptionalPassthrough
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.transformations.series.difference import Differencer
from sktime.transformations.series.boxcox import LogTransformer
from sktime.forecasting.base import ForecastingHorizon
from sktime.datasets import load_airline

y = load_airline()
X = pd.DataFrame({"lag1": y.shift(1).fillna(method="bfill"),
                  "lag2": y.shift(2).fillna(method="bfill")}, 
                 index=y.index)

y_train, y_test, X_train, X_test = temporal_train_test_split(y=y,X=X, test_size=30)
fh_out = np.arange(1, len(y_test)+1)
fh_ins = ForecastingHorizon(y_train.index, is_relative=False)

pipe = ForecastingPipeline(
    steps=[
        (
            "transformersX",
            TransformerPipeline(
                steps=[
                    ("logX", OptionalPassthrough(LogTransformer())),
                    ("differencerX", Differencer(na_handling="fill_zero")),
                ]
            ),
        ),
        (
            "forecaster",
            TransformedTargetForecaster(
                steps=[
                    (
                        "transformers",
                        TransformerPipeline(
                            steps=[
                                ("log", OptionalPassthrough(LogTransformer())),
                                ("differencer", Differencer(na_handling="fill_zero")),
                            ]
                        ),
                    ),
                    ("arima", ARIMA()),
                ]
            ),
        ),
    ]
)

pipe.fit(y=y_train, X=X_train)
pipe_ins = pipe.predict(fh=fh_ins, X=X_train)
pipe_forecast = pipe.predict(fh=fh_out, X=X_test)

arima_model = ARIMA()
arima_model.fit(y=y_train, X=X_train)
model_ins = arima_model.predict(fh=fh_ins, X=X_train)
model_forecast = arima_model.predict(fh=fh_out, X=X_test)


# Returns True - not expected
np.allclose(y_train.to_numpy(), pipe_ins.to_numpy())

# Returns False - not expected
np.allclose(pipe_ins.to_numpy(), model_ins.to_numpy())

# Returns False - expected
np.allclose(y_train.to_numpy(), model_ins.to_numpy())
```

**Comment by ilkersigirci** [comment, marked as answer]
I guess #3346 fixes this problem. Thanks.
